core-bundle/src/Image/Studio/FigureBuilder.php

```diff
@@ -0,0 +1,469 @@
+
+        return $this;
+    }
+
+    /**
+     * Set a custom light box group id. By default or if the argument is set to
+     * null an id will be generated.
+     *
+     * For this setting to take effect make sure you enabled the creation of a
+     * light box by calling `enableLightBox()`.
+     */
+    public function setLightBoxGroupIdentifier(?string $identifier): self
+    {
+        $this->lightBoxGroupIdentifier = $identifier;
+
+        return $this;
+    }
+
+    /**
+     * Enable/disable creation of a light box image (if possible) and/or
+     * outputting the respective link attributes. This setting is disabled by
+     * default.
+     */
+    public function enableLightBox(bool $enable = true): self
+    {
+        $this->enableLightBox = $enable;
+
+        return $this;
+    }
+
+    /**
+     * Create a result object with the current settings.
+     */
+    public function build(): Figure
+    {
+        if (null === $this->filePath) {
+            throw new \LogicException('You need to set a resource before building the result.');
+        }
+
+        // Freeze settings to allow reusing this builder object.
+        $settings = clone $this;
+
+        $imageResult = $this->locator
+            ->get('contao.image.studio')
+            ->createImage($settings->filePath, $this->sizeConfiguration)
+        ;
+
+        // We're defining some values via a Closure to make their evaluation lazy
+        return new Figure(
+            $imageResult,
+            function (Figure $figure) use ($settings): ?MetaData {
+                return $this->onDefineMetaData($settings);
+            },
+            $settings->additionalLinkAttributes,
+            function (Figure $figure) use ($settings): ?LightBoxResult {
+                return $this->onDefineLightBoxResult($settings, $figure);
+            }
+        );
+    }
+
+    /**
+     * Define meta data [on demand].
+     */
+    private function onDefineMetaData(self $settings): ?MetaData
+    {
+        if (null !== $settings->metaData) {
+            return $settings->metaData;
+        }
+
+        if (null === $settings->filesModel) {
+            return null;
+        }
+
+        // Get fallback locale list or use without fallbacks if explicitly set
+        $locales = null !== $settings->locale ? [$settings->locale] : $this->getFallbackLocaleList();
+
+        return $settings->filesModel->getMetaData(...$locales);
+    }
+
+    /**
+     * Define the light box result if it is enabled [on demand].
+     */
+    private function onDefineLightBoxResult(self $settings, Figure $result): ?LightBoxResult
+    {
+        if (!$settings->enableLightBox) {
+            return null;
+        }
+
+        // Use explicitly set uri (1), fall back to using meta data (2) or use the base resource (3) if empty.
+        $lightBoxUri = $settings->lightBoxUri ?? Controller::replaceInsertTags($result->getMetaData()->getUrl()) ?: $settings->filePath;
```

Comment:
Is this the correct place for `Controller::replaceInsertTags()`? How about replacing the insert tags before passing the data to the MetaData object?

Reply:
Yes, I thought about that as well.. I wonder if you would you ever want to have the intact insert tag in the output?

Reply:
If the insert-tag should be cached differently, this might make sense (E.g. `© {{date::Y}} Contao`). But for the URL-field we can IMO replace them directly in the Models when creating the MetaData object.